// webusb/lib/adapter.js
"use strict";
/*
* Node WebUSB
* Copyright (c) 2017 Rob Moran
*
* The MIT License (MIT)
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var usb_1 = require("usb");
var configuration_1 = require("./configuration");
var interface_1 = require("./interface");
var alternate_1 = require("./alternate");
var endpoint_1 = require("./endpoint");
var device_1 = require("./device");
/**
 * @hidden
 */
var DEFAULT_DELAY_TIMEOUT = 200;
/**
 * @hidden
 */
var DEFAULT_RETRY_COUNT = 10;
/**
 * @hidden
 */
var CONSTANTS = {
    WEB_UUID: "3408b638-09a9-47a0-8bfd-a0768815b665",
    LIBUSB_DT_BOS: 0x0f,
    LIBUSB_DT_BOS_SIZE: 0x05,
    LIBUSB_TRANSFER_TYPE_MASK: 0x03,
    USB_VERSION: 0x201,
    CAPABILITY_VERSION: 0x0100,
    URL_REQUEST_TYPE: 0xC0,
    URL_REQUEST_INDEX: 0x02,
    CLEAR_FEATURE: 0x01,
    ENDPOINT_HALT: 0x00
};
/**
 * @hidden
 */
var USBAdapter = /** @class */ (function (_super) {
    __extends(USBAdapter, _super);
    function USBAdapter() {
        var _this = _super.call(this) || this;
        // Maintains a live list of connected Web USB devices
        _this.devices = {};
        var attachCallback = function (device) {
            _this.loadDevice(device, DEFAULT_RETRY_COUNT)
                .then(function (loadedDevice) {
                if (loadedDevice) {
                    var handle = _this.getDeviceHandle(loadedDevice);
                    _this.devicetoUSBDevice(handle)
                        .then(function (usbDevice) {
                        if (usbDevice) {
                            _this.emit(USBAdapter.EVENT_DEVICE_CONNECT, usbDevice);
                        }
                    });
                }
            });
        };
        var detachCallback = function (device) {
            var handle = _this.getDeviceHandle(device);
            if (handle && _this.devices[handle]) {
                delete _this.devices[handle];
                _this.emit(USBAdapter.EVENT_DEVICE_DISCONNECT, handle);
            }
        };
        _this.on("newListener", function (event) {
            var listenerCount = _this.listenerCount(event);
            if (listenerCount !== 0) {
                return;
            }
            if (event === USBAdapter.EVENT_DEVICE_CONNECT) {
                usb_1.on("attach", attachCallback);
            }
            else if (event === USBAdapter.EVENT_DEVICE_DISCONNECT) {
                usb_1.on("detach", detachCallback);
            }
        });
        _this.on("removeListener", function (event) {
            var listenerCount = _this.listenerCount(event);
            if (listenerCount !== 0) {
                return;
            }
            if (event === USBAdapter.EVENT_DEVICE_CONNECT) {
                usb_1.removeListener("attach", attachCallback);
            }
            else if (event === USBAdapter.EVENT_DEVICE_DISCONNECT) {
                usb_1.removeListener("detach", detachCallback);
            }
        });
        return _this;
    }
    USBAdapter.prototype.getDeviceHandle = function (device) {
        if (device.busNumber === null || device.deviceAddress === null) {
            return null;
        }
        return device.busNumber + "." + device.deviceAddress;
    };
    USBAdapter.prototype.serialPromises = function (task, params) {
        function reducer(chain, param) {
            var _this = this;
            return chain
                .then(function (results) {
                return task.call(_this, param)
                    .then(function (result) {
                    if (result) {
                        results.push(result);
                    }
                    return results;
                });
            });
        }
        return params.reduce(reducer.bind(this), Promise.resolve([]));
    };
    USBAdapter.prototype.serialDevicePromises = function (task, device, descriptors) {
        function reducer(chain, descriptor) {
            var _this = this;
            return chain
                .then(function (results) {
                return task.call(_this, device, descriptor)
                    .then(function (result) {
                    results.push(result);
                    return results;
                });
            });
        }
        return descriptors.reduce(reducer.bind(this), Promise.resolve([]));
    };
    USBAdapter.prototype.delay = function (timeout) {
        if (timeout === void 0) { timeout = DEFAULT_DELAY_TIMEOUT; }
        return new Promise(function (resolve, _reject) {
            setTimeout(resolve, timeout);
        });
    };
    USBAdapter.prototype.retryPromise = function (fn, retries, timeout) {
        var _this = this;
        if (retries === void 0) { retries = 0; }
        if (timeout === void 0) { timeout = DEFAULT_DELAY_TIMEOUT; }
        return new Promise(function (resolve, reject) {
            fn()
                .then(resolve)
                .catch(function (error) {
                if (retries === 0) {
                    return reject(error);
                }
                return _this.delay(timeout)
                    .then(function () { return _this.retryPromise(fn, --retries, timeout); })
                    .then(resolve);
            });
        });
    };
    USBAdapter.prototype.loadDevices = function () {
        var devices = usb_1.getDeviceList();
        return this.serialPromises(this.loadDevice, devices);
    };
    USBAdapter.prototype.loadDevice = function (device, retries) {
        var _this = this;
        if (retries === void 0) { retries = 0; }
        // Early guard against unsupported USB devices
        try {
            // tslint:disable-next-line:no-unused-expression
            device.configDescriptor;
            // tslint:disable-next-line:no-unused-expression
            device.allConfigDescriptors;
            // tslint:disable-next-line:no-unused-expression
            device.deviceDescriptor;
        }
        catch (_error) {
            return Promise.resolve(null);
        }
        return this.getCapabilities(device, retries)
            .then(function (capabilities) { return _this.getWebCapability(capabilities); })
            .then(function (capability) {
            return _this.getWebUrl(device, capability)
                .then(function (url) {
                var handle = _this.getDeviceHandle(device);
                _this.devices[handle] = {
                    device: device,
                    url: url
                };
                return device;
            });
        });
    };
    USBAdapter.prototype.getCapabilities = function (device, retries) {
        var _this = this;
        return new Promise(function (resolve, _reject) {
            _this.openDevice(device, retries)
                .then(function () {
                device.getCapabilities(function (error, capabilities) {
                    try {
                        // Older macs (<10.12) can error with some host devices during a close at this point
                        device.close();
                        // tslint:disable-next-line:no-empty
                    }
                    catch (_error) { }
                    if (error)
                        return resolve([]);
                    resolve(capabilities);
                });
            })
                .catch(function (_error) {
                resolve([]);
            });
        });
    };
    USBAdapter.prototype.getWebCapability = function (capabilities) {
        var _this = this;
        var platformCapabilities = capabilities.filter(function (capability) {
            return capability.type === 5;
        });
        var webCapability = platformCapabilities.find(function (capability) {
            var uuid = _this.decodeUUID(capability.data.slice(1, 17));
            var version = capability.data.readUInt16LE(17);
            return uuid === CONSTANTS.WEB_UUID && version === CONSTANTS.CAPABILITY_VERSION;
        });
        return webCapability;
    };
    USBAdapter.prototype.decodeUUID = function (buffer) {
        var data1 = ("00000000" + buffer.readUInt32LE(0).toString(16)).slice(-8);
        var data2 = ("0000" + buffer.readUInt16LE(4).toString(16)).slice(-4);
        var data3 = ("0000" + buffer.readUInt16LE(6).toString(16)).slice(-4);
        var data4 = [];
        for (var i = 8; i < 10; i++) {
            data4.push(("00" + buffer.readUInt8(i).toString(16)).slice(-2));
        }
        var data5 = [];
        for (var i = 10; i < 16; i++) {
            data5.push(("00" + buffer.readUInt8(i).toString(16)).slice(-2));
        }
        return data1 + "-" + data2 + "-" + data3 + "-" + data4.join("") + "-" + data5.join("");
    };
    USBAdapter.prototype.getWebUrl = function (device, capability, suppressErrors) {
        var _this = this;
        if (suppressErrors === void 0) { suppressErrors = true; }
        return new Promise(function (resolve, reject) {
            if (!capability || !capability.data || capability.data.byteLength < 20)
                return resolve(null);
            var vendor = capability.data.readUInt8(19);
            var page = capability.data.readUInt8(20);
            _this.openDevice(device)
                .then(function () {
                device.controlTransfer(CONSTANTS.URL_REQUEST_TYPE, vendor, page, CONSTANTS.URL_REQUEST_INDEX, 64, function (error, buffer) {
                    device.close();
                    if (error) {
                        // An error may be due to the URL not existing
                        if (suppressErrors)
                            return resolve(null);
                        else
                            return reject(error);
                    }
                    // const length = buffer.readUInt8(0);
                    // const type = buffer.readUInt8(1);
                    var url = buffer.toString("utf8", 3);
                    var scheme = buffer.readUInt8(2); // 0 - http, 1 - https, 255 - in url
                    if (scheme === 0)
                        url = "http://" + url;
                    if (scheme === 1)
                        url = "https://" + url;
                    resolve(url);
                });
            })
                .catch(function (_error) {
                resolve("");
            });
        });
    };
    USBAdapter.prototype.devicetoUSBDevice = function (handle) {
        var _this = this;
        return new Promise(function (resolve, _reject) {
            var device = _this.devices[handle].device;
            var url = _this.devices[handle].url;
            var configs = null;
            var configDescriptor = null;
            var deviceDescriptor = null;
            try {
                configDescriptor = device.configDescriptor;
                configs = device.allConfigDescriptors;
                deviceDescriptor = device.deviceDescriptor;
            }
            catch (_error) {
                return resolve(null);
            }
            if (!configs)
                return resolve(null);
            return _this.serialDevicePromises(_this.configToUSBConfiguration, device, configs)
                .then(function (configurations) {
                if (!deviceDescriptor) {
                    return resolve(new device_1.USBDevice({
                        _handle: _this.getDeviceHandle(device),
                        url: url,
                        configurations: configurations
                    }));
                }
                var deviceVersion = _this.decodeVersion(deviceDescriptor.bcdDevice);
                var usbVersion = _this.decodeVersion(deviceDescriptor.bcdUSB);
                var manufacturerName = null;
                var productName = null;
                return _this.getStringDescriptor(device, deviceDescriptor.iManufacturer)
                    .then(function (name) {
                    manufacturerName = name;
                    return _this.getStringDescriptor(device, deviceDescriptor.iProduct);
                })
                    .then(function (name) {
                    productName = name;
                    return _this.getStringDescriptor(device, deviceDescriptor.iSerialNumber);
                })
                    .then(function (serialNumber) {
                    var props = {
                        _handle: _this.getDeviceHandle(device),
                        _maxPacketSize: deviceDescriptor.bMaxPacketSize0,
                        url: url,
                        deviceClass: deviceDescriptor.bDeviceClass,
                        deviceSubclass: deviceDescriptor.bDeviceSubClass,
                        deviceProtocol: deviceDescriptor.bDeviceProtocol,
                        productId: deviceDescriptor.idProduct,
                        vendorId: deviceDescriptor.idVendor,
                        deviceVersionMajor: deviceVersion.major,
                        deviceVersionMinor: deviceVersion.minor,
                        deviceVersionSubminor: deviceVersion.sub,
                        usbVersionMajor: usbVersion.major,
                        usbVersionMinor: usbVersion.minor,
                        usbVersionSubminor: usbVersion.sub,
                        manufacturerName: manufacturerName,
                        productName: productName,
                        serialNumber: serialNumber,
                        configurations: configurations,
                        _currentConfiguration: configDescriptor.bConfigurationValue
                    };
                    return resolve(new device_1.USBDevice(props));
                });
            }).catch(function (_error) {
                resolve(null);
            });
        });
    };
    USBAdapter.prototype.decodeVersion = function (version) {
        var hex = ("0000" + version.toString(16)).slice(-4);
        return {
            major: parseInt(hex.substr(0, 2), null),
            minor: parseInt(hex.substr(2, 1), null),
            sub: parseInt(hex.substr(3, 1), null),
        };
    };
    USBAdapter.prototype.getStringDescriptor = function (device, index) {
        var _this = this;
        return new Promise(function (resolve) {
            _this.openDevice(device)
                .then(function () {
                device.getStringDescriptor(index, function (error, buffer) {
                    device.close();
                    resolve(error ? "" : buffer.toString());
                });
            })
                .catch(function (_error) {
                resolve("");
            });
        });
    };
    USBAdapter.prototype.bufferToDataView = function (buffer) {
        var arrayBuffer = new Uint8Array(buffer).buffer;
        return new DataView(arrayBuffer);
    };
    USBAdapter.prototype.bufferSourceToBuffer = function (bufferSource) {
        var arrayBuffer = ArrayBuffer.isView(bufferSource)
            ? bufferSource.buffer.slice(bufferSource.byteOffset, bufferSource.byteOffset + bufferSource.byteLength)
            : bufferSource;
        return new Buffer(arrayBuffer);
    };
    USBAdapter.prototype.getEndpoint = function (device, direction, endpointNumber) {
        var endpoint = null;
        var address = endpointNumber | (direction === "in" ? usb_1.LIBUSB_ENDPOINT_IN : usb_1.LIBUSB_ENDPOINT_OUT);
        device.interfaces.some(function (iface) {
            var epoint = iface.endpoint(address);
            if (epoint) {
                endpoint = epoint;
                return true;
            }
        });
        return endpoint;
    };
    USBAdapter.prototype.getInEndpoint = function (device, endpointNumber) {
        var endpoint = this.getEndpoint(device, "in", endpointNumber);
        if (endpoint && endpoint.direction === "in")
            return endpoint;
    };
    USBAdapter.prototype.getOutEndpoint = function (device, endpointNumber) {
        var endpoint = this.getEndpoint(device, "out", endpointNumber);
        if (endpoint && endpoint.direction === "out")
            return endpoint;
    };
    USBAdapter.prototype.endpointToUSBEndpoint = function (descriptor) {
        var direction = descriptor.bEndpointAddress & usb_1.LIBUSB_ENDPOINT_IN ? "in" : "out";
        return new endpoint_1.USBEndpoint({
            endpointNumber: descriptor.bEndpointAddress ^ (direction === "in" ? usb_1.LIBUSB_ENDPOINT_IN : usb_1.LIBUSB_ENDPOINT_OUT),
            direction: direction,
            type: (descriptor.bmAttributes & CONSTANTS.LIBUSB_TRANSFER_TYPE_MASK) === usb_1.LIBUSB_TRANSFER_TYPE_BULK ? "bulk"
                : (descriptor.bmAttributes & CONSTANTS.LIBUSB_TRANSFER_TYPE_MASK) === usb_1.LIBUSB_TRANSFER_TYPE_INTERRUPT ? "interrupt"
                    : "isochronous",
            packetSize: descriptor.wMaxPacketSize
        });
    };
    USBAdapter.prototype.interfaceToUSBAlternateInterface = function (device, descriptor) {
        var _this = this;
        return this.getStringDescriptor(device, descriptor.iInterface)
            .then(function (name) {
            return new alternate_1.USBAlternateInterface({
                alternateSetting: descriptor.bAlternateSetting,
                interfaceClass: descriptor.bInterfaceClass,
                interfaceSubclass: descriptor.bInterfaceSubClass,
                interfaceProtocol: descriptor.bInterfaceProtocol,
                interfaceName: name,
                endpoints: descriptor.endpoints.map(_this.endpointToUSBEndpoint)
            });
        });
    };
    USBAdapter.prototype.interfacesToUSBInterface = function (device, descriptors) {
        var _this = this;
        return this.serialDevicePromises(this.interfaceToUSBAlternateInterface, device, descriptors)
            .then(function (alternates) {
            return new interface_1.USBInterface({
                _handle: _this.getDeviceHandle(device),
                interfaceNumber: descriptors[0].bInterfaceNumber,
                alternates: alternates
            });
        });
    };
    USBAdapter.prototype.configToUSBConfiguration = function (device, descriptor) {
        var _this = this;
        return this.getStringDescriptor(device, descriptor.iConfiguration)
            .then(function (name) {
            var allInterfaces = descriptor.interfaces || [];
            return _this.serialDevicePromises(_this.interfacesToUSBInterface, device, allInterfaces)
                .then(function (interfaces) {
                return new configuration_1.USBConfiguration({
                    configurationValue: descriptor.bConfigurationValue,
                    configurationName: name,
                    interfaces: interfaces
                });
            });
        });
    };
    USBAdapter.prototype.getDevice = function (handle) {
        if (!this.devices[handle])
            return null;
        return this.devices[handle].device;
    };
    USBAdapter.prototype.controlTransferParamsToType = function (setup, direction) {
        var recipient = setup.recipient === "device" ? usb_1.LIBUSB_RECIPIENT_DEVICE
            : setup.recipient === "interface" ? usb_1.LIBUSB_RECIPIENT_INTERFACE
                : setup.recipient === "endpoint" ? usb_1.LIBUSB_RECIPIENT_ENDPOINT
                    : usb_1.LIBUSB_RECIPIENT_OTHER;
        var requestType = setup.requestType === "standard" ? usb_1.LIBUSB_REQUEST_TYPE_STANDARD
            : setup.requestType === "class" ? usb_1.LIBUSB_REQUEST_TYPE_CLASS
                : usb_1.LIBUSB_REQUEST_TYPE_VENDOR;
        return recipient | requestType | direction;
    };
    USBAdapter.prototype.openDevice = function (device, retries) {
        if (retries === void 0) { retries = 0; }
        return this.retryPromise(function () {
            return new Promise(function (resolve, reject) {
                try {
                    device.open();
                }
                catch (error) {
                    return reject(error);
                }
                resolve();
            });
        }, retries);
    };
    USBAdapter.prototype.getConnected = function (handle) {
        return this.getDevice(handle) !== null;
    };
    USBAdapter.prototype.getOpened = function (handle) {
        var device = this.getDevice(handle);
        if (!device)
            return false;
        return (device.interfaces !== null);
    };
    USBAdapter.prototype.listUSBDevices = function () {
        var _this = this;
        return this.loadDevices()
            .then(function () {
            return _this.serialPromises(_this.devicetoUSBDevice, Object.keys(_this.devices));
        });
    };
    USBAdapter.prototype.open = function (handle) {
        var device = this.getDevice(handle);
        return this.openDevice(device);
    };
    USBAdapter.prototype.close = function (handle) {
        var _this = this;
        return new Promise(function (resolve, _reject) {
            var device = _this.getDevice(handle);
            device.close();
            resolve();
        });
    };
    USBAdapter.prototype.selectConfiguration = function (handle, id) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var device = _this.getDevice(handle);
            device.setConfiguration(id, function (error) {
                if (error)
                    return reject(error);
                resolve();
            });
        });
    };
    USBAdapter.prototype.claimInterface = function (handle, address) {
        var _this = this;
        return new Promise(function (resolve, _reject) {
            var device = _this.getDevice(handle);
            device.interface(address).claim();
            resolve();
        });
    };
    USBAdapter.prototype.releaseInterface = function (handle, address) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var device = _this.getDevice(handle);
            device.interface(address).release(true, function (error) {
                if (error)
                    return reject(error);
                resolve();
            });
        });
    };
    USBAdapter.prototype.selectAlternateInterface = function (handle, interfaceNumber, alternateSetting) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var device = _this.getDevice(handle);
            var iface = device.interface(interfaceNumber);
            iface.setAltSetting(alternateSetting, function (error) {
                if (error)
                    return reject(error);
                resolve();
            });
        });
    };
    USBAdapter.prototype.controlTransferIn = function (handle, setup, length) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var device = _this.getDevice(handle);
            var type = _this.controlTransferParamsToType(setup, usb_1.LIBUSB_ENDPOINT_IN);
            device.controlTransfer(type, setup.request, setup.value, setup.index, length, function (error, buffer) {
                if (error) {
                    if (error.errno === usb_1.LIBUSB_TRANSFER_STALL) {
                        return resolve({
                            status: "stall"
                        });
                    }
                    else if (error.errno === usb_1.LIBUSB_TRANSFER_OVERFLOW) {
                        return resolve({
                            status: "babble"
                        });
                    }
                    return reject(error);
                }
                resolve({
                    data: _this.bufferToDataView(buffer),
                    status: "ok"
                });
            });
        });
    };
    USBAdapter.prototype.controlTransferOut = function (handle, setup, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var device = _this.getDevice(handle);
            var type = _this.controlTransferParamsToType(setup, usb_1.LIBUSB_ENDPOINT_OUT);
            var buffer = data ? _this.bufferSourceToBuffer(data) : new Buffer(0);
            device.controlTransfer(type, setup.request, setup.value, setup.index, buffer, function (error) {
                if (error) {
                    if (error.errno === usb_1.LIBUSB_TRANSFER_STALL) {
                        return resolve({
                            bytesWritten: 0,
                            status: "stall"
                        });
                    }
                    return reject(error);
                }
                resolve({
                    bytesWritten: buffer.byteLength,
                    status: "ok" // hack
                });
            });
        });
    };
    USBAdapter.prototype.clearHalt = function (handle, direction, endpointNumber) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var device = _this.getDevice(handle);
            var wIndex = endpointNumber | (direction === "in" ? usb_1.LIBUSB_ENDPOINT_IN : usb_1.LIBUSB_ENDPOINT_OUT);
            device.controlTransfer(usb_1.LIBUSB_RECIPIENT_ENDPOINT, CONSTANTS.CLEAR_FEATURE, CONSTANTS.ENDPOINT_HALT, wIndex, 0, function (error) {
                if (error)
                    return reject(error);
                resolve();
            });
        });
    };
    USBAdapter.prototype.transferIn = function (handle, endpointNumber, length) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var device = _this.getDevice(handle);
            var endpoint = _this.getInEndpoint(device, endpointNumber);
            endpoint.transfer(length, function (error, data) {
                if (error) {
                    if (error.errno === usb_1.LIBUSB_TRANSFER_STALL) {
                        return resolve({
                            status: "stall"
                        });
                    }
                    else if (error.errno === usb_1.LIBUSB_TRANSFER_OVERFLOW) {
                        return resolve({
                            status: "babble"
                        });
                    }
                    return reject(error);
                }
                resolve({
                    data: _this.bufferToDataView(data),
                    status: "ok"
                });
            });
        });
    };
    USBAdapter.prototype.transferOut = function (handle, endpointNumber, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var device = _this.getDevice(handle);
            var endpoint = _this.getOutEndpoint(device, endpointNumber);
            var buffer = _this.bufferSourceToBuffer(data);
            endpoint.transfer(buffer, function (error) {
                if (error) {
                    if (error.errno === usb_1.LIBUSB_TRANSFER_STALL) {
                        return resolve({
                            bytesWritten: 0,
                            status: "stall"
                        });
                    }
                    return reject(error);
                }
                resolve({
                    bytesWritten: buffer.byteLength,
                    status: "ok" // hack
                });
            });
        });
    };
    USBAdapter.prototype.isochronousTransferIn = function (_handle, _endpointNumber, _packetLengths) {
        return new Promise(function (_resolve, reject) {
            reject("isochronousTransferIn error: method not implemented");
        });
    };
    USBAdapter.prototype.isochronousTransferOut = function (_handle, _endpointNumber, _data, _packetLengths) {
        return new Promise(function (_resolve, reject) {
            reject("isochronousTransferOut error: method not implemented");
        });
    };
    USBAdapter.prototype.reset = function (handle) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var device = _this.getDevice(handle);
            device.reset(function (error) {
                if (error)
                    return reject(error);
                resolve();
            });
        });
    };
    USBAdapter.EVENT_DEVICE_CONNECT = "connect";
    USBAdapter.EVENT_DEVICE_DISCONNECT = "disconnect";
    return USBAdapter;
}(events_1.EventEmitter));
exports.USBAdapter = USBAdapter;
/**
 * @hidden
 */
exports.adapter = new USBAdapter();
