// webusb/lib/device.js
"use strict";
/*
* Node WebUSB
* Copyright (c) 2017 Rob Moran
*
* The MIT License (MIT)
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/
Object.defineProperty(exports, "__esModule", { value: true });
var usb_1 = require("usb");
var adapter_1 = require("./adapter");
/**
 * USB Device
 */
var USBDevice = /** @class */ (function () {
    /**
     * @hidden
     */
    function USBDevice(init) {
        /**
         * Manufacturer name of the device
         */
        this.manufacturerName = null;
        /**
         * Product name of the device
         */
        this.productName = null;
        /**
         * Serial number of the device
         */
        this.serialNumber = null;
        this._configurations = [];
        /**
         * @hidden
         */
        this._currentConfiguration = null;
        /**
         * URL advertised by the device (not part of Web USB specification)
         */
        this.url = null;
        /**
         * @hidden
         */
        this._maxPacketSize = 0;
        /**
         * @hidden
         */
        this._handle = null;
        this.usbVersionMajor = init.usbVersionMajor;
        this.usbVersionMinor = init.usbVersionMinor;
        this.usbVersionSubminor = init.usbVersionSubminor;
        this.deviceClass = init.deviceClass;
        this.deviceSubclass = init.deviceSubclass;
        this.deviceProtocol = init.deviceProtocol;
        this.vendorId = init.vendorId;
        this.productId = init.productId;
        this.deviceVersionMajor = init.deviceVersionMajor;
        this.deviceVersionMinor = init.deviceVersionMinor;
        this.deviceVersionSubminor = init.deviceVersionSubminor;
        this.manufacturerName = init.manufacturerName;
        this.productName = init.productName;
        this.serialNumber = init.serialNumber;
        this._configurations = init.configurations;
        this.url = init.url;
        this._maxPacketSize = init._maxPacketSize;
        this._handle = init._handle;
        this._currentConfiguration = init._currentConfiguration;
    }
    Object.defineProperty(USBDevice.prototype, "configurations", {
        /**
         * List of configurations supported by the device
         */
        get: function () {
            return this._configurations;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(USBDevice.prototype, "configuration", {
        /**
         * The currently selected configuration
         */
        get: function () {
            var _this = this;
            return this.configurations.find(function (configuration) { return configuration.configurationValue === _this._currentConfiguration; });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(USBDevice.prototype, "connected", {
        /**
         * @hidden
         */
        get: function () {
            return adapter_1.adapter.getConnected(this._handle);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(USBDevice.prototype, "opened", {
        /**
         * A flag indicating whether the device is open
         */
        get: function () {
            return adapter_1.adapter.getOpened(this._handle);
        },
        enumerable: true,
        configurable: true
    });
    USBDevice.prototype.getEndpoint = function (direction, endpointNumber) {
        var endpoint = null;
        var iface = null;
        this.configuration.interfaces.some(function (usbInterface) {
            endpoint = usbInterface.alternate.endpoints.find(function (usbEndpoint) {
                return (usbEndpoint.endpointNumber === endpointNumber && usbEndpoint.direction === direction);
            });
            if (endpoint)
                iface = usbInterface;
            return endpoint;
        });
        return {
            endpoint: endpoint,
            iface: iface
        };
    };
    USBDevice.prototype.setupInvalid = function (setup) {
        if (setup.recipient === "interface") {
            var interfaceNumber_1 = setup.index & 0xff; // lower 8 bits
            var iface = this.configuration.interfaces.find(function (usbInterface) { return usbInterface.interfaceNumber === interfaceNumber_1; });
            if (!iface)
                return "interface not found";
            if (!iface.claimed)
                return "invalid state";
        }
        else if (setup.recipient === "endpoint") {
            var endpointNumber = setup.index & 0x0f; // lower 4 bits
            var direction = setup.index & usb_1.LIBUSB_ENDPOINT_IN ? "in" : "out";
            var result = this.getEndpoint(direction, endpointNumber);
            if (!result.endpoint)
                return "endpoint not found";
            if (!result.iface.claimed)
                return "invalid state";
        }
    };
    /**
     * Opens the device
     */
    USBDevice.prototype.open = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("open error: device not found"));
            if (_this.opened)
                return resolve();
            adapter_1.adapter.open(_this._handle)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("open error: " + error));
            });
        });
    };
    /**
     * Closes the device
     */
    USBDevice.prototype.close = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("close error: device not found"));
            if (!_this.opened)
                return resolve();
            adapter_1.adapter.close(_this._handle)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("close error: " + error));
            });
        });
    };
    /**
     * Select a configuration for the device
     * @param configurationValue The configuration value to select
     * @returns Promise containing any error
     */
    USBDevice.prototype.selectConfiguration = function (configurationValue) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            // Don't change the configuration if it's already set correctly
            if (configurationValue === _this._currentConfiguration)
                return resolve();
            if (!_this.connected)
                return reject(new Error("selectConfiguration error: device not found"));
            var config = _this.configurations.find(function (configuration) { return configuration.configurationValue === configurationValue; });
            if (!config)
                return reject(new Error("selectConfiguration error: configuration not found"));
            if (!_this.opened)
                return reject(new Error("selectConfiguration error: invalid state"));
            adapter_1.adapter.selectConfiguration(_this._handle, configurationValue)
                .then(function () {
                _this._currentConfiguration = configurationValue;
                _this.configuration.interfaces.forEach(function (iface) { return iface.reset(); });
                resolve();
            })
                .catch(function (error) {
                reject(new Error("selectConfiguration error: " + error));
            });
        });
    };
    /**
     * Claim an interface on the device
     * @param interfaceNumber The interface number to claim
     * @returns Promise containing any error
     */
    USBDevice.prototype.claimInterface = function (interfaceNumber) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("claimInterface error: device not found"));
            var iface = _this.configuration.interfaces.find(function (usbInterface) { return usbInterface.interfaceNumber === interfaceNumber; });
            if (!iface)
                return reject(new Error("claimInterface error: interface not found"));
            if (!_this.opened)
                return reject(new Error("claimInterface error: invalid state"));
            if (iface.claimed)
                return resolve();
            iface.claimInterface()
                .then(resolve)
                .catch(function (error) {
                reject(new Error("claimInterface error: " + error));
            });
        });
    };
    /**
     * Release an interface on the device
     * @param interfaceNumber The interface number to release
     * @returns Promise containing any error
     */
    USBDevice.prototype.releaseInterface = function (interfaceNumber) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("releaseInterface error: device not found"));
            var iface = _this.configuration.interfaces.find(function (usbInterface) { return usbInterface.interfaceNumber === interfaceNumber; });
            if (!iface)
                return reject(new Error("releaseInterface error: interface not found"));
            if (!_this.opened)
                return reject(new Error("releaseInterface error: invalid state"));
            if (!iface.claimed)
                return resolve();
            iface.releaseInterface()
                .then(resolve)
                .catch(function (error) {
                reject(new Error("releaseInterface error: " + error));
            });
        });
    };
    /**
     * Select an alternate interface on the device
     * @param interfaceNumber The interface number to change
     * @param alternateSetting The alternate setting to use
     * @returns Promise containing any error
     */
    USBDevice.prototype.selectAlternateInterface = function (interfaceNumber, alternateSetting) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("selectAlternateInterface error: device not found"));
            var iface = _this.configuration.interfaces.find(function (usbInterface) { return usbInterface.interfaceNumber === interfaceNumber; });
            if (!iface)
                return reject(new Error("selectAlternateInterface error: interface not found"));
            if (!_this.opened || !iface.claimed)
                return reject(new Error("selectAlternateInterface error: invalid state"));
            iface.selectAlternateInterface(alternateSetting)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("selectAlternateInterface error: " + error));
            });
        });
    };
    /**
     * Undertake a control transfer in from the device
     *
     * @param setup The USB control transfer parameters
     * @param length The amount of data to transfer
     * @returns Promise containing a result
     */
    USBDevice.prototype.controlTransferIn = function (setup, length) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("controlTransferIn error: device not found"));
            if (!_this.opened)
                return reject(new Error("controlTransferIn error: invalid state"));
            var setupError = _this.setupInvalid(setup);
            if (setupError)
                return reject(new Error("controlTransferIn error: " + setupError));
            adapter_1.adapter.controlTransferIn(_this._handle, setup, length)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("controlTransferIn error: " + error));
            });
        });
    };
    /**
     * Undertake a control transfer out to the device
     *
     * __Note:__ The bytesWritten always set to the length of the data
     *
     * @param setup The USB control transfer parameters
     * @param data The data to transfer
     * @returns Promise containing a result
     */
    USBDevice.prototype.controlTransferOut = function (setup, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("controlTransferOut error: device not found"));
            if (!_this.opened)
                return reject(new Error("controlTransferOut error: invalid state"));
            var setupError = _this.setupInvalid(setup);
            if (setupError)
                return reject(new Error("controlTransferOut error: " + setupError));
            adapter_1.adapter.controlTransferOut(_this._handle, setup, data)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("controlTransferOut error: " + error));
            });
        });
    };
    /**
     * Clear a halt condition on an endpoint
     *
     * @param direction The direction of the endpoint to clear
     * @param endpointNumber The endpoint number of the endpoint to clear
     * @returns Promise containing any error
     */
    USBDevice.prototype.clearHalt = function (direction, endpointNumber) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("clearHalt error: device not found"));
            var result = _this.getEndpoint(direction, endpointNumber);
            if (!result.endpoint)
                return reject(new Error("clearHalt error: endpoint not found"));
            if (!_this.opened || !result.iface.claimed)
                return reject(new Error("clearHalt error: invalid state"));
            adapter_1.adapter.clearHalt(_this._handle, direction, endpointNumber)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("clearHalt error: " + error));
            });
        });
    };
    /**
     * Undertake a transfer in from the device
     *
     * @param endpointNumber The number of the endpoint to transfer from
     * @param length The amount of data to transfer
     * @returns Promise containing a result
     */
    USBDevice.prototype.transferIn = function (endpointNumber, length) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("transferIn error: device not found"));
            var result = _this.getEndpoint("in", endpointNumber);
            if (!result.endpoint)
                return reject(new Error("transferIn error: endpoint not found"));
            if (result.endpoint.type !== "interrupt" && result.endpoint.type !== "bulk")
                return reject(new Error("transferIn error: invalid access"));
            if (!_this.opened || !result.iface.claimed)
                return reject(new Error("transferIn error: invalid state"));
            adapter_1.adapter.transferIn(_this._handle, endpointNumber, length)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("transferIn error: " + error));
            });
        });
    };
    /**
     * Undertake a transfer out to the device
     *
     * __Note:__ The bytesWritten always set to the length of the data
     *
     * @param endpointNumber The number of the endpoint to transfer to
     * @param data The data to transfer
     * @returns Promise containing a result
     */
    USBDevice.prototype.transferOut = function (endpointNumber, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("transferOut error: device not found"));
            var result = _this.getEndpoint("out", endpointNumber);
            if (!result.endpoint)
                return reject(new Error("transferOut error: endpoint not found"));
            if (result.endpoint.type !== "interrupt" && result.endpoint.type !== "bulk")
                return reject(new Error("transferOut error: invalid access"));
            if (!_this.opened || !result.iface.claimed)
                return reject(new Error("transferOut error: invalid state"));
            adapter_1.adapter.transferOut(_this._handle, endpointNumber, data)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("transferOut error: " + error));
            });
        });
    };
    /**
     * @hidden
     * Undertake an isochronous transfer in from the device
     * @param endpointNumber The number of the endpoint to transfer from
     * @param packetLengths An array of packet lengths outlining the amount to transfer
     * @returns Promise containing a result
     */
    USBDevice.prototype.isochronousTransferIn = function (endpointNumber, packetLengths) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("isochronousTransferIn error: device not found"));
            var result = _this.getEndpoint("in", endpointNumber);
            if (!result.endpoint)
                return reject(new Error("isochronousTransferIn error: endpoint not found"));
            if (result.endpoint.type !== "isochronous")
                return reject(new Error("isochronousTransferIn error: invalid access"));
            if (!_this.opened || !result.iface.claimed)
                return reject(new Error("isochronousTransferIn error: invalid state"));
            adapter_1.adapter.isochronousTransferIn(_this._handle, endpointNumber, packetLengths)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("isochronousTransferIn error: " + error));
            });
        });
    };
    /**
     * @hidden
     * Undertake an isochronous transfer out to the device
     * @param endpointNumber The number of the endpoint to transfer to
     * @param data The data to transfer
     * @param packetLengths An array of packet lengths outlining the amount to transfer
     * @returns Promise containing a result
     */
    USBDevice.prototype.isochronousTransferOut = function (endpointNumber, data, packetLengths) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("isochronousTransferOut error: device not found"));
            var result = _this.getEndpoint("out", endpointNumber);
            if (!result.endpoint)
                return reject(new Error("isochronousTransferOut error: endpoint not found"));
            if (result.endpoint.type !== "isochronous")
                return reject(new Error("isochronousTransferOut error: invalid access"));
            if (!_this.opened || !result.iface.claimed)
                return reject(new Error("isochronousTransferOut error: invalid state"));
            adapter_1.adapter.isochronousTransferOut(_this._handle, endpointNumber, data, packetLengths)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("isochronousTransferOut error: " + error));
            });
        });
    };
    /**
     * Soft reset the device
     * @returns Promise containing any error
     */
    USBDevice.prototype.reset = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.connected)
                return reject(new Error("reset error: device not found"));
            if (!_this.opened)
                return reject(new Error("reset error: invalid state"));
            adapter_1.adapter.reset(_this._handle)
                .then(resolve)
                .catch(function (error) {
                reject(new Error("reset error: " + error));
            });
        });
    };
    return USBDevice;
}());
exports.USBDevice = USBDevice;
