// webusb/lib/usb.js
"use strict";
/*
* Node WebUSB
* Copyright (c) 2017 Rob Moran
*
* The MIT License (MIT)
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var dispatcher_1 = require("./dispatcher");
var adapter_1 = require("./adapter");
/**
 * USB class
 */
var USB = /** @class */ (function (_super) {
    __extends(USB, _super);
    /**
     * USB constructor
     * @param options USB initialisation options
     */
    function USB(options) {
        var _this = _super.call(this) || this;
        _this.allowedDevices = [];
        _this.devicesFound = null;
        options = options || {};
        _this.devicesFound = options.devicesFound;
        var deviceConnectCallback = function (device) {
            if (_this.replaceAllowedDevice(device)) {
                _this.emit(USB.EVENT_DEVICE_CONNECT, device);
            }
        };
        var deviceDisconnectCallback = function (handle) {
            var allowedDevice = _this.allowedDevices.find(function (allowedDevices) { return allowedDevices._handle === handle; });
            if (allowedDevice) {
                _this.emit(USB.EVENT_DEVICE_DISCONNECT, allowedDevice);
            }
        };
        _this.on("newListener", function (event) {
            var listenerCount = _this.listenerCount(event);
            if (listenerCount !== 0) {
                return;
            }
            if (event === USB.EVENT_DEVICE_CONNECT) {
                adapter_1.adapter.addListener(adapter_1.USBAdapter.EVENT_DEVICE_CONNECT, deviceConnectCallback);
            }
            else if (event === USB.EVENT_DEVICE_DISCONNECT) {
                adapter_1.adapter.addListener(adapter_1.USBAdapter.EVENT_DEVICE_DISCONNECT, deviceDisconnectCallback);
            }
        });
        _this.on("removeListener", function (event) {
            var listenerCount = _this.listenerCount(event);
            if (listenerCount !== 0) {
                return;
            }
            if (event === USB.EVENT_DEVICE_CONNECT) {
                adapter_1.adapter.removeListener(adapter_1.USBAdapter.EVENT_DEVICE_CONNECT, deviceConnectCallback);
            }
            else if (event === USB.EVENT_DEVICE_DISCONNECT) {
                adapter_1.adapter.removeListener(adapter_1.USBAdapter.EVENT_DEVICE_DISCONNECT, deviceDisconnectCallback);
            }
        });
        return _this;
    }
    USB.prototype.replaceAllowedDevice = function (device) {
        for (var i in this.allowedDevices) {
            if (this.allowedDevices[i].productId === device.productId
                && this.allowedDevices[i].vendorId === device.vendorId
                && this.allowedDevices[i].serialNumber === device.serialNumber) {
                this.allowedDevices[i] = device;
                return true;
            }
        }
        return false;
    };
    USB.prototype.filterDevice = function (options, device) {
        return options.filters.some(function (filter) {
            // Vendor
            if (filter.vendorId && filter.vendorId !== device.vendorId)
                return false;
            // Product
            if (filter.productId && filter.productId !== device.productId)
                return false;
            // Class
            if (filter.classCode) {
                // Interface Descriptors
                var match = device.configuration.interfaces.some(function (iface) {
                    // Class
                    if (filter.classCode && filter.classCode !== iface.alternate.interfaceClass)
                        return false;
                    // Subclass
                    if (filter.subclassCode && filter.subclassCode !== iface.alternate.interfaceSubclass)
                        return false;
                    // Protocol
                    if (filter.protocolCode && filter.protocolCode !== iface.alternate.interfaceProtocol)
                        return false;
                    return true;
                });
                if (match)
                    return true;
            }
            // Class
            if (filter.classCode && filter.classCode !== device.deviceClass)
                return false;
            // Subclass
            if (filter.subclassCode && filter.subclassCode !== device.deviceSubclass)
                return false;
            // Protocol
            if (filter.protocolCode && filter.protocolCode !== device.deviceProtocol)
                return false;
            // Serial
            if (filter.serialnumber && filter.serialnumber !== device.serialNumber)
                return false;
            return true;
        });
    };
    /**
     * Gets all allowed Web USB devices
     * @returns Promise containing an array of devices
     */
    USB.prototype.getDevices = function () {
        var _this = this;
        return new Promise(function (resolve, _reject) {
            resolve(_this.allowedDevices);
        });
    };
    /**
     * Requests a single Web USB device
     * @param options The options to use when scanning
     * @returns Promise containing the selected device
     */
    USB.prototype.requestDevice = function (options) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            // Must have options
            if (!options) {
                return reject(new TypeError("requestDevice error: 1 argument required, but only 0 present"));
            }
            // Options must be an object
            if (options.constructor !== {}.constructor) {
                return reject(new TypeError("requestDevice error: parameter 1 (options) is not an object"));
            }
            // Must have a filter
            if (!options.filters) {
                return reject(new TypeError("requestDevice error: required member filters is undefined"));
            }
            // Filter must be an array
            if (options.filters.constructor !== [].constructor) {
                return reject(new TypeError("requestDevice error: the provided value cannot be converted to a sequence"));
            }
            // Check filters
            var check = options.filters.every(function (filter) {
                // Protocol & Subclass
                if (filter.protocolCode && !filter.subclassCode) {
                    reject(new TypeError("requestDevice error: subclass code is required"));
                    return false;
                }
                // Subclass & Class
                if (filter.subclassCode && !filter.classCode) {
                    reject(new TypeError("requestDevice error: class code is required"));
                    return false;
                }
                return true;
            });
            if (!check)
                return;
            return adapter_1.adapter.listUSBDevices()
                .then(function (devices) {
                devices = devices.filter(function (device) { return _this.filterDevice(options, device); });
                if (devices.length === 0) {
                    return reject(new Error("requestDevice error: no devices found"));
                }
                function selectFn(device) {
                    if (!this.replaceAllowedDevice(device))
                        this.allowedDevices.push(device);
                    resolve(device);
                }
                // If no devicesFound function, select the first device found
                if (!_this.devicesFound)
                    return selectFn.call(_this, devices[0]);
                var selectedDevice = _this.devicesFound(devices, selectFn.bind(_this));
                if (selectedDevice)
                    selectFn.call(_this, selectedDevice);
            }).catch(function (error) {
                reject(new Error("requestDevice error: " + error));
            });
        });
    };
    /**
     * Allowed device Connected event
     * @event
     */
    USB.EVENT_DEVICE_CONNECT = "connect";
    /**
     * Allowed device Disconnected event
     * @event
     */
    USB.EVENT_DEVICE_DISCONNECT = "disconnect";
    return USB;
}(dispatcher_1.EventDispatcher));
exports.USB = USB;
